blockio-sim/src/System/FS/BlockIO/Sim.hs

```diff
@@ -32,22 +35,64 @@ fromHasFS hfs = serialHasBlockIO hSetNoCache hAdvise hAllocate tryLockFile hfs
     hAdvise _ _ _ _ = pure ()
     hAllocate _ _ _ = pure ()
 
-    -- We simulate file locks by opening files in write-mode (exclusive lock) or
-    -- read-mode (shared lock). If files are opened in conflicting modes, i.e.,
-    -- write+write or read+write, then hOpen will throw an FsError, which is
-    -- interpreted as the file locking having failed.
-    --
-    -- This simulation does not simulate inter-process locking. In fact, the
-    -- simulation only works properly if all the tested code uses the same
-    -- simHasFS/simHasBlockIO instances. That is the case not just for file
-    -- locks, but in general for most if not all HasFS/HasBlockIO operations.
-    tryLockFile :: FsPath -> LockMode -> m (Maybe (LockFileHandle m))
-    tryLockFile path ExclusiveLock = flip catch (\FsError{} -> pure Nothing) $ do
-      h <- hOpen hfs path (WriteMode AllowExisting)
-      pure (Just (LockFileHandle (hClose hfs h)))
-    tryLockFile path SharedLock = flip catch (\FsError{} -> pure Nothing) $ do
-      h <- hOpen hfs path ReadMode
-      pure (Just (LockFileHandle (hClose hfs h)))
+-- | Lock files are reader\/writer locks.
+--
+-- We implement this using the content of the lock file. The content is a
+-- counter, positive for readers and negaive (specifically -1) for writers.
+-- There can be any number of readers, but only one writer.
+--
+-- Warning: This implementation is not robust under concurrent use (because
+-- operations on files are not atomic) but should be ok for casual use. A
+-- proper implementation would need to be part of the underlying 'HasFs'
+-- implementations.
+--
+simTryLockFile :: forall m h.
+                  MonadThrow m
+               => HasFS m h
+               -> FsPath
+               -> LockMode
+               -> m (Maybe (LockFileHandle m))
+simTryLockFile hfs path lockmode =
+    API.withFile hfs path (ReadWriteMode AllowExisting) $ \h -> do
+      n <- readCount h
+      case lockmode of
+        SharedLock    | n >= 0 -> do writeCount h (n+1)
+                                     return (Just LockFileHandle { hUnlock })
+        ExclusiveLock | n == 0 -> do writeCount h (-1)
+                                     return (Just LockFileHandle { hUnlock })
+        _                      -> return Nothing
+  where
+    hUnlock =
+      API.withFile hfs path (ReadWriteMode AllowExisting) $ \h -> do
+        n <- readCount h
+        case lockmode of
+          SharedLock    | n >  0  -> writeCount h (n-1)
+          ExclusiveLock | n == -1 -> writeCount h 0
+          _                       -> throwIO countCorrupt
+
+    readCount :: Handle h -> m Int
+    readCount h = do
+      content <- API.hGetSome hfs h 10
```

Comment:
We should use `hGetAllAt` or something similar, because `hGetSome` can do partial reads